vllm/scripts.py

```diff
@@ -24,7 +28,49 @@ def signal_handler(sig, frame):
     signal.signal(signal.SIGTSTP, signal_handler)
 
 
+def _merge_args_and_config(args: argparse.Namespace):
+    """
+        merge args from cli and config file supplied in the cli. 
+        If an argument is present in cli and args then choose args value
+        over config file.
+
+        example:
+        # config.yaml
+        port: 1231
+
+        # invoke server
+        $ vllm serve --config ../config.yaml --port 3122
+
+        # selected port = 3122
+    """
+    assert args.config, 'No config file specified.'
+
+    # only expecting a flat dictionary of atomic types
+    config: Dict[str, Union[int, str]] = {}
+
+    try:
+        with open(args.config, 'r') as config_file:
+            config = yaml.safe_load(config_file)
+    except Exception as ex:
+        logger.error("Unable to read the config file at %s", args.config)
+        logger.error(ex)
+
+    for key, value in config.items():
+        if hasattr(args, key):
```

Comment:
how does this deal with options that have default values?

Reply:
tested. same behaviour.  default args are populated in the args.Namespace before it enters the method.

Reply:
*same ebhaviour as in cli takes presidence, if specified in config file as well. and default in case no where speficied..

Reply:
I mean, even if I don't specify `-tp 1`, `args.tensor_parallel` will have value. in this case, the default cli value takes precedence over config file, which is not desirable.

Reply:
please bear with me @youkaichao 

i see so if i understand correctly.. ideal behavior would be that; the user supplies ALLthe args in config file, if --config has been used, even the default ones? In a manner of speaking the config file will always have ~105 lines in them (current population of namespace)..
so for example:

Ex1

```yaml
#config.yaml

port: 1234
```
```python
$ vllm serve "fabebook/opt-12b" --config config.yaml
$ [ERROR] Please specify all args in config file while using --config 
```

Ex2

```yaml
#config.yaml

port: 1234
... 104 more..
```
```python
$ vllm serve "fabebook/opt-12b" --config config.yaml --port 4
$ [INFO] port = 4
```

Ex3

```python
$ vllm serve "fabebook/opt-12b" 
$ [INFO] port = 8000 (default)
```

Reply:
Looks like a misunderstanding.

I'm trying to understand, how do you know if `getattr(args, key)` is a default value or a value from commandline?

Reply:
oooh got it! 
I understand it now. we will have to handle other cases as well with validating if arguments in config file have membership with the vllm namespace, shorthand etc and c.. 
with this in mind i think its prudent for us to use 1configargparse1 which inherits from 1argparse1.
this will keep the diff small with largest ROI as well. moreover, it doesnt neeed anything more than what is already in vllm's requirements.

Reply:
PS:
I have also tested the the scratch implementation using sys.argv check for default membership and then overriding args appropirately. we can go ahead with that too if its too much of a leap in faith to use configargparse module..

Reply:
@youkaichao  hope you are doing well. 

Unfortunately with `configargparse` the tests are not passing. The server is not able to start quickly enough for the api tests to pass..

I will go ahead with the sys.argv implementation since its "lightweight" and more tailor made for our purposes.

Reply:
@youkaichao  i feel good  about this. Please let me know what do you think.